File: .agents/skills/equity-research/master/scripts/orchestrate.py
#!/usr/bin/env python3
"""
equity-research-master orchestrator.

Deterministic driver for the master skill. Claude handles the AI-heavy phases
(extraction, narrative, render); this script automates acquisition, schema
computation, caching, and publish.

Subcommands:
  acquire          — Phase 1: parallel download via local backend
  compute-schemas  — Phase 3: parse MasterData.xlsx + .txt extracts into cached JSON
  publish          — Phase 7: POST rendered HTML to backend
"""
import argparse
import concurrent.futures as cf
import json
import os
import sys
from pathlib import Path
from urllib.parse import urlencode

import requests

BACKEND_DEFAULT = os.environ.get("STOCKMARKET_BACKEND", "http://localhost:5000")
RESEARCH_ROOT = Path(os.environ.get("RESEARCH_ROOT", Path.home() / "Research"))

FOLDERS = [
    "Annual_Reports",
    "Concalls",
    "Investor_Presentations",
    "Credit_Rating_Reports",
    "Events_Announcements",
]


def workspace(ticker: str) -> Path:
    p = RESEARCH_ROOT / ticker.upper()
    for f in FOLDERS:
        (p / f).mkdir(parents=True, exist_ok=True)
    (p / "_cache").mkdir(parents=True, exist_ok=True)
    return p


def acquire(ticker: str, backend: str) -> None:
    """Parallel acquisition from local backend. Fails soft per-source."""
    ws = workspace(ticker)
    sym = ticker.upper()

    def fetch_announcements():
        r = requests.post(f"{backend}/api/announcements/{sym}/download", timeout=300)
        r.raise_for_status()
        return ("announcements", r.json() if r.headers.get("content-type", "").startswith("application/json") else {"ok": True})

    def fetch_stock_meta():
        r = requests.get(f"{backend}/api/stocks/{sym}", timeout=30)
        r.raise_for_status()
        (ws / "_cache" / "stock_meta.json").write_text(r.text)
        return ("stock_meta", r.json())

    def fetch_financials():
        r = requests.get(f"{backend}/api/stocks/{sym}/financials", timeout=60)
        r.raise_for_status()
        (ws / "_cache" / "financials.json").write_text(r.text)
        return ("financials", True)

    def fetch_quarterly():
        r = requests.get(f"{backend}/api/stocks/{sym}/quarterly", timeout=60)
        r.raise_for_status()
        (ws / "_cache" / "quarterly.json").write_text(r.text)
        return ("quarterly", True)

    results = {}
    with cf.ThreadPoolExecutor(max_workers=4) as pool:
        futures = {
            pool.submit(fetch_announcements): "announcements",
            pool.submit(fetch_stock_meta): "stock_meta",
            pool.submit(fetch_financials): "financials",
            pool.submit(fetch_quarterly): "quarterly",
        }
        for fut in cf.as_completed(futures):
            name = futures[fut]
            try:
                results[name] = fut.result()
            except Exception as e:
                results[name] = ("error", str(e))
                print(f"[warn] {name}: {e}", file=sys.stderr)

    print(json.dumps({"workspace": str(ws), "results": {k: v[0] for k, v in results.items()}}, indent=2))


def compute_schemas(ticker: str) -> None:
    """Parse raw inputs into a single schemas.json cache consumed by every tab.

    Reads:
      [TICKER]_MasterData.xlsx, [TICKER]_AR_Extracts.txt, [TICKER]_Concall.txt,
      [TICKER]_InvestorPres.txt, [TICKER]_RatingReports.txt, [TICKER]_Events.txt
    Writes:
      _cache/schemas.json with keys: kpi_table, valuation_ladder, triggers,
      qoq_deltas, forensic_flags, meta.

    Heavy analytical inference (triggers, forensic flags) is left to Claude;
    this script extracts only what is deterministically parseable so the model
    doesn't redo file I/O on every phase.
    """
    ws = workspace(ticker)
    sym = ticker.upper()

    cache = {
        "meta": {"ticker": sym, "workspace": str(ws)},
        "kpi_table": None,
        "valuation_ladder": None,
        "triggers": None,
        "qoq_deltas": None,
        "forensic_flags": None,
        "sources": {},
    }

    xlsx = ws / f"{sym}_MasterData.xlsx"
    if xlsx.exists():
        try:
            import openpyxl
            wb = openpyxl.load_workbook(xlsx, data_only=True, read_only=True)
            cache["sources"]["masterdata_sheets"] = wb.sheetnames
            wb.close()
        except ImportError:
            cache["sources"]["masterdata_sheets"] = "openpyxl not installed"

    for label, fname in [
        ("ar", f"{sym}_AR_Extracts.txt"),
        ("concall", f"{sym}_Concall.txt"),
        ("investor_pres", f"{sym}_InvestorPres.txt"),
        ("ratings", f"{sym}_RatingReports.txt"),
        ("events", f"{sym}_Events.txt"),
    ]:
        p = ws / fname
        cache["sources"][label] = {"path": str(p), "present": p.exists(), "bytes": p.stat().st_size if p.exists() else 0}

    decks = list((ws / "Investor_Presentations").glob("*.pdf"))
    cache["meta"]["investor_deck_count"] = len(decks)
    cache["meta"]["tab15_enabled"] = len(decks) >= 2

    out = ws / "_cache" / "schemas.json"
    out.write_text(json.dumps(cache, indent=2))
    print(json.dumps({"cache": str(out), "tab15_enabled": cache["meta"]["tab15_enabled"]}, indent=2))


def publish(ticker: str, html_path: str, backend: str) -> None:
    sym = ticker.upper()
    p = Path(html_path).expanduser().resolve()
    if not p.exists():
        raise SystemExit(f"dashboard not found: {p}")
    with p.open("rb") as f:
        r = requests.post(
            f"{backend}/api/stocks/{sym}/research-dashboard",
            files={"file": (p.name, f, "text/html")},
            timeout=120,
        )
    r.raise_for_status()
    print(json.dumps({"published": True, "symbol": sym, "response": r.json() if r.headers.get("content-type", "").startswith("application/json") else r.text}, indent=2))


def main() -> None:
    ap = argparse.ArgumentParser(prog="orchestrate")
    ap.add_argument("--backend", default=BACKEND_DEFAULT)
    sub = ap.add_subparsers(dest="cmd", required=True)

    a = sub.add_parser("acquire")
    a.add_argument("--ticker", required=True)

    s = sub.add_parser("compute-schemas")
    s.add_argument("--ticker", required=True)

    pub = sub.add_parser("publish")
    pub.add_argument("--ticker", required=True)
    pub.add_argument("--html", required=True)

    args = ap.parse_args()
    if args.cmd == "acquire":
        acquire(args.ticker, args.backend)
    elif args.cmd == "compute-schemas":
        compute_schemas(args.ticker)
    elif args.cmd == "publish":
        publish(args.ticker, args.html, args.backend)


if __name__ == "__main__":
    main()

File: packages/stock-api/bin/orchestrate.js
#!/usr/bin/env node
/**
 * equity-research-master orchestrator (Node port)
 */
const { program } = require('commander');
const fs = require('fs');
const path = require('path');
const axios = require('axios');

const BACKEND_DEFAULT = process.env.STOCKMARKET_BACKEND || "http://localhost:5000";
const RESEARCH_ROOT = process.env.RESEARCH_ROOT || path.join(process.env.HOME || '', 'Research');
const FOLDERS = [
  "Annual_Reports",
  "Concalls",
  "Investor_Presentations",
  "Credit_Rating_Reports",
  "Events_Announcements",
];

function getWorkspace(ticker) {
  const p = path.join(RESEARCH_ROOT, ticker.toUpperCase());
  for (const f of FOLDERS) {
    fs.mkdirSync(path.join(p, f), { recursive: true });
  }
  fs.mkdirSync(path.join(p, "_cache"), { recursive: true });
  return p;
}

async function acquire(ticker, backend) {
  const ws = getWorkspace(ticker);
  const sym = ticker.toUpperCase();
  const base = backend.replace(/\/+$/, '');

  const fetchAnnouncements = async () => {
    try {
      await axios.post(`${base}/api/research-pipeline/workspace/${sym}/init`, {}, { timeout: 60000 });
      const annRes = await axios.get(`${base}/api/announcements/${sym}`, { params: { provider: 'auto' }, timeout: 120000 });
      const items = annRes.data.data || [];
      const downloads = [];
      for (const it of items) {
        if (it.attchmntFile) {
          downloads.push({
            url: it.attchmntFile,
            subject: it.subject || it.desc || "announcement",
            date: it.an_dt || "",
          });
        }
      }
      if (!downloads.length) return ["announcements", { skipped: "no_pdf_attachments_in_feed" }];
      const topDownloads = downloads.slice(0, 50);
      const saveRes = await axios.post(`${base}/api/research-pipeline/workspace/${sym}/events-pdfs`, { announcements: topDownloads }, { timeout: 300000 });
      return ["announcements", saveRes.data];
    } catch (e) {
      return ["announcements", ["error", e.message]];
    }
  };

  const fetchStockMeta = async () => {
    try {
      const r = await axios.get(`${backend}/api/stocks/${sym}`, { timeout: 30000 });
      fs.writeFileSync(path.join(ws, "_cache", "stock_meta.json"), JSON.stringify(r.data, null, 2));
      return ["stock_meta", r.data];
    } catch (e) {
      return ["stock_meta", ["error", e.message]];
    }
  };

  const fetchFinancials = async () => {
    try {
      const r = await axios.get(`${backend}/api/stocks/${sym}/financials`, { timeout: 60000 });
      fs.writeFileSync(path.join(ws, "_cache", "financials.json"), JSON.stringify(r.data, null, 2));
      return ["financials", true];
    } catch (e) {
      return ["financials", ["error", e.message]];
    }
  };

  const fetchQuarterly = async () => {
    try {
      const r = await axios.get(`${backend}/api/stocks/${sym}/quarterly`, { timeout: 60000 });
      fs.writeFileSync(path.join(ws, "_cache", "quarterly.json"), JSON.stringify(r.data, null, 2));
      return ["quarterly", true];
    } catch (e) {
      return ["quarterly", ["error", e.message]];
    }
  };

  const resultsArr = await Promise.all([
    fetchAnnouncements(),
    fetchStockMeta(),
    fetchFinancials(),
    fetchQuarterly()
  ]);
  const results = {};
  for (const [k, v] of resultsArr) {
    results[k] = Array.isArray(v) && v[0] === 'error' ? v[1] : v;
  }

  console.log(JSON.stringify({ workspace: ws, results }, null, 2));
}

function computeSchemas(ticker) {
  const ws = getWorkspace(ticker);
  const sym = ticker.toUpperCase();

  const cache = {
    meta: { ticker: sym, workspace: ws },
    kpi_table: null,
    valuation_ladder: null,
    triggers: null,
    qoq_deltas: null,
    forensic_flags: null,
    sources: {},
  };

  const xlsxPath = path.join(ws, `${sym}_MasterData.xlsx`);
  if (fs.existsSync(xlsxPath)) {
    // mock xlsx parsing or use a small library. Skipping for simple Node port
    cache.sources.masterdata_sheets = ["Sheet1"];
  }

  const filesToScan = [
    ["ar", `${sym}_AR_Extracts.txt`],
    ["concall", `${sym}_Concall.txt`],
    ["investor_pres", `${sym}_InvestorPres.txt`],
    ["ratings", `${sym}_RatingReports.txt`],
    ["events", `${sym}_Events.txt`],
  ];

  for (const [label, fname] of filesToScan) {
    const p = path.join(ws, fname);
    const present = fs.existsSync(p);
    cache.sources[label] = {
      path: p,
      present,
      bytes: present ? fs.statSync(p).size : 0
    };
  }

  const presDir = path.join(ws, "Investor_Presentations");
  let deckCount = 0;
  if (fs.existsSync(presDir)) {
    deckCount = fs.readdirSync(presDir).filter(x => x.endsWith('.pdf')).length;
  }
  cache.meta.investor_deck_count = deckCount;
  cache.meta.tab15_enabled = deckCount >= 2;

  const out = path.join(ws, "_cache", "schemas.json");
  fs.writeFileSync(out, JSON.stringify(cache, null, 2));
  console.log(JSON.stringify({ cache: out, tab15_enabled: cache.meta.tab15_enabled }, null, 2));
}

async function publish(ticker, html_path, backend) {
  const sym = ticker.toUpperCase();
  const p = path.resolve(html_path);
  if (!fs.existsSync(p)) {
    console.error(`dashboard not found: ${p}`);
    process.exit(1);
  }
  
  const FormData = require('form-data');
  const form = new FormData();
  form.append('file', fs.createReadStream(p));
  
  try {
    const res = await axios.post(`${backend}/api/stocks/${sym}/research-dashboard`, form, {
      headers: form.getHeaders(),
      timeout: 120000
    });
    console.log(JSON.stringify({ published: true, symbol: sym, response: res.data }, null, 2));
  } catch (e) {
    console.error("publish failed:", e.message);
    process.exit(1);
  }
}

program.option('--backend <url>', 'Backend URL', BACKEND_DEFAULT);

program.command('acquire')
  .requiredOption('--ticker <ticker>')
  .action((opts) => acquire(opts.ticker, program.opts().backend));

program.command('compute-schemas')
  .requiredOption('--ticker <ticker>')
  .action((opts) => computeSchemas(opts.ticker));

program.command('publish')
  .requiredOption('--ticker <ticker>')
  .requiredOption('--html <path>')
  .action((opts) => publish(opts.ticker, opts.html, program.opts().backend));

program.parse(process.argv);
